src/plugins/discover/public/application/angular/doc_table/components/table_row.tsx

```diff
@@ -0,0 +1,258 @@
+    [services]
+  );
+
+  // toggle display of the rows details, a full list of the fields from each row
+  const toggleRow = () => setOpen((prevOpen) => !prevOpen);
+
+  /**
+   * Fill an element with the value of a field
+   */
+  const _displayField = useCallback(
+    (fieldName: string, truncate = false) => {
+      const text = indexPattern.formatField(row, fieldName);
+      // eslint-disable-next-line react/no-danger
+      const fieldElement = <span dangerouslySetInnerHTML={{ __html: text }} />;
+
+      if (truncate && text.length > MIN_LINE_LENGTH) {
+        return <div className="truncate-by-height">{fieldElement}</div>;
+      }
+
+      return fieldElement;
+    },
+    [indexPattern, row]
+  );
+
+  const inlineFilter = (column: string, type: '+' | '-') => {
+    const field = indexPattern.fields.getByName(column);
+    filter(field!, flattenedRow[column], type);
+  };
+
+  const getContextAppHref = () => {
+    return getContextUrl(
+      row._id,
+      indexPattern.id!,
+      columns,
+      services.filterManager,
+      services.addBasePath
+    );
+  };
+
+  const getSingleDocHref = () => {
+    return services.addBasePath(
+      `/app/discover#/doc/${indexPattern.id!}/${row._index}?id=${encodeURIComponent(row._id)}`
+    );
+  };
+
+  const newHtmls = [
+    <td className="kbnDocTableCell__toggleDetails" key="toggleDetailsCell">
+      <EuiButtonEmpty
+        onClick={toggleRow}
+        size="xs"
+        aria-expanded={!!open}
+        aria-label={i18n.translate('discover.docTable.tableRow.toggleRowDetailsButtonAriaLabel', {
+          defaultMessage: 'Toggle row details',
+        })}
+        data-test-subj="docTableExpandToggleColumn"
+      >
+        {open && <EuiIcon type="arrowDown" size="s" />}
+        {!open && <EuiIcon type="arrowRight" size="s" />}
+      </EuiButtonEmpty>
+    </td>,
+  ];
+
+  if (indexPattern.timeFieldName && !hideTimeColumn) {
+    newHtmls.push(
+      <TableCell
+        key={indexPattern.timeFieldName}
+        timefield={true}
+        formatted={_displayField(indexPattern.timeFieldName)}
+        filterable={Boolean(mapping(indexPattern.timeFieldName)?.filterable && filter)}
+        column={indexPattern.timeFieldName}
+        inlineFilter={inlineFilter}
+      />
+    );
+  }
+
+  if (columns.length === 0 && useNewFieldsApi) {
+    const formatted = formatRow(row, indexPattern);
+
+    newHtmls.push(
+      <TableCell
+        key="__document__"
+        timefield={false}
+        sourcefield={true}
+        formatted={formatted}
+        filterable={false}
+        column="__document__"
+        inlineFilter={inlineFilter}
+      />
+    );
+  } else {
+    columns.forEach(function (column: string) {
+      const isFilterable = Boolean(mapping(column) && mapping(column)?.filterable && filter);
+      if (useNewFieldsApi && !mapping(column) && !row.fields![column]) {
+        const innerColumns = Object.fromEntries(
+          Object.entries(row.fields!).filter(([key]) => {
```

Comment:
This type casting is safe here due to when `useNewFieldsApi` is true `fields` property will be defained

Reply:
Could you please put that as an inline comment into the code, so we have this reason later documented in code :)

Reply:
Done